README.md
# cr-mindTree

react component

## Example
 <img src="./github/example1.png" alt="example" />

## Usage

```jsx
cnpm install --save cr-mindtree
```

```js
import MindTree from 'cr-mindtree'
import 'cr-mindtree/assets/index.css'
// 注意在使用的时候不能忘记引入css

<MindTree title='分支机构' bgColor='#e4effc'
  tree={[{
    name: 'leaf-on0',
    value: '001',
    accessory: '75%',
  }, {
    name: 'leaf-two',
    value: '002',
    accessory: '25%',
  }]}
  mode='bottomCenter'
/>
```

## 注意
如果使用了cssModule，作为样式的使用方式，在引入css时应特别注意
以And Design Pro使用为例，在对应的页面less中引入css。注意包名前面的波浪线
```less
@import '~cr-mindtree/assets/index.css';
```
## API
name | type | description
-----|------|------------
title|string|组件标题
tree|array|传入组件树数据，类似[{name: 'leaf-one', value: '001'}, {name: 'leaf-two', value: '002'}]
bgColor|string|组件背景颜色
mode   |string|组件展开的方式（'bottomCenter', 'topRight', 'topLeft', 'bottomLeft', 'bottomRight'）
treeChange|func|组件展开收起 function(val) {}; 组件展开val为true,收起为false
leafClick|func|点击组件叶子事件 function(event, val) {}; val为叶子数据
treeBoxWidth|number|组件树叶子部分的宽度，一般不需要设置，默认值为300
className|string|为组建增加className
hoverBgColor|string|组件枝叶及标题hover背景颜色
## Development

```
cnpm install
npm start
```

## Example

http://localhost:8019/examples/testOne.html

## Test
```js
npm test
```


## License

cr-mindTree is released under the MIT license.


HISTORY.md
# History

## 1.10.3
增加 zf-tools 功能

## 1.10.2
组件枝叶文本过长无法显示...bug

## 1.10.0
增加组件枝叶及标题hover背景颜色，有需要可自行添加

## 1.0.9
优化枝叶文本过长显示问题，做加省略号加title处理

## 1.0.8
优化只有一个数据项时的显示

## 1.0.7
去掉枝叶默认白色背景问题，以及优化枝叶线图片

## 1.0.6 / 2019-06-06
修复className 不传值为 undefined 问题和 className 无法使用 cssModule 模式的问题
修复tree参数不传值时有默认Dom问题

## 1.0.5 / 2019-06-06
增加className property

## 1.0.4 / 2019-06-04
修改top模式，切换时数组自动反转问题
修改切换动画，只能上滑不能下滑问题

## 1.0.3 / 2019-06-03
增加枝叶的附属项

## 1.0.0 / 2019-05-29

cr-mindTree is start


tests/index.spec.js
import React from 'react'
import { render, mount, shallow } from 'enzyme'
import { renderToJson } from 'enzyme-to-json'
import MindTree from '../index'
import expect from 'expect.js'

describe('MindTree basic', () => {
  it('renders correctly', () => {
    const wrapper = mount(<MindTree bgColor='#6eabec' title='南京云帐房网络科技有限公司' />);
    expect(wrapper.props().title).to.equal('南京云帐房网络科技有限公司');
  })

  it('render tree props', () => {
    const wrapper = render(
      <MindTree 
        title='分支机构'
        tree={[{
          name: '测试一',
          value: 'ce-shi-yi'
        }, {
          name: '测试二',
          value: 'ce-shi-er'
        }]}
      />
    );
    expect(wrapper.find('.cr-mind-tree-tree-list').length).to.equal(2)
  })

  it('check the click event', () => {
    const wrapper = mount(
      <MindTree 
        title='分支机构'
        tree={[{
          name: '测试一',
          value: 'ce-shi-yi'
        }, {
          name: '测试二',
          value: 'ce-shi-er'
        }]}
      />
    )
    wrapper.find('.cr-mind-tree-title').simulate('click')
    expect(wrapper.find('.anticon').hasClass('anticon-plus-circle'))
  })
  
})

examples/testOne.js
import React from 'react'
import ReactDOM from 'react-dom'
import PropTypes from 'prop-types'
import MindTree from '../index'
import '../assets/index.less'
import '../assets/example.less'


const reactContainer = document.getElementById('__react-content');
const bodyContainer = document.getElementsByTagName('body')
bodyContainer[0].style.padding = '10px'

function handleTreeChange(val) {
  console.log(val)
}
function handleLeafClick(e, value) {
  console.log(value)
}

function render(container) {
  ReactDOM.render(
    <div>
      <h3>MINDTREE TEST</h3>
      <div>
      <h3>无枝叶模式</h3>
      <MindTree className={"treeTitle"} bgColor='#6eabec' title='南京云帐房网络科技有限公司' />
      </div>
      <div style={{minHeight: '300px'}}>
      <h3>模式一：bottomCenter</h3>
        <MindTree title='分支机构' bgColor='#e4effc'
          tree={[{
            name: '刘丹',
            value: 'liu-dan'
          }, {
            name: 'hello',
            value: 'hello',        
          }]}
          mode='bottomCenter'
          treeChange={handleTreeChange}
          leafClick={handleLeafClick}
        />
      </div>
      <div style={{minHeight: '300px', marginLeft: '300px'}}>
      <h3 style={{marginLeft: '-100px'}}>模式二：topRight</h3>
        <MindTree title='分支机构' bgColor='#e4effc' 
          tree={[{
            name: '刘丹000',
            value: 'liu-dan',
            accessory: '12%',
          }, {
            name: '李易峰测试文本长度如果很长会有问题呢',
            value: 'li-yi-feng',
            accessory: '88%',          
          }]}
          mode='topRight'
          treeChange={handleTreeChange}
          leafClick={handleLeafClick}
          treeBoxWidth={300}
        />
      </div>
      <div style={{minHeight: '500px'}}>
      <h3>模式三：bottomRight</h3>
        <MindTree title='分支机构' bgColor='#e4effc'
          tree={[{
            name: '刘丹',
            value: 'liu-dan',
          }, {
            name: '李易峰测试是文本长达李易峰测试是文本长达',
            value: 'li-yi-feng',          
          }]}
          mode='bottomRight'
          treeChange={handleTreeChange}
          leafClick={handleLeafClick}
        />
      </div>   
      <div style={{minHeight: '300px', marginLeft: '300px'}}>
      <h3>模式四：topLeft</h3>
        <MindTree title='分支机构' bgColor='red'
          tree={[{
            name: '刘丹',
            value: 'liu-dan',
            accessory: '20%',
          }, {
            name: '李易峰454666',
            value: 'li-yi-feng',
            accessory: '88%',           
          }]}
          mode='topLeft'
          treeChange={handleTreeChange}
          leafClick={handleLeafClick}
          hoverBgColor='green'
        />
      </div>   
      <div style={{minHeight: '300px', marginLeft: '300px'}}>
      <h3>模式五：bottomLeft</h3>
        <MindTree title='分支机构' bgColor='#e4effc'
          tree={[{
            name: '刘丹',
            value: 'liu-dan',
            accessory: '董事',
          }, {
            name: '李易峰',
            value: 'li-yi-feng',
            accessory: '88%',           
          }]}
          mode='bottomLeft'
          treeChange={handleTreeChange}
          leafClick={handleLeafClick}
        />
      </div>                 
    </div>, container
  )
}

render(reactContainer)

src/index.js
import React, { PureComponent } from 'react'
import PropTypes from 'prop-types'
import OuiDom from './utils/ouiDomUtils'
import { Icon } from 'antd'
import velocity from 'velocity-animate'
import domAlign from 'dom-align'
import classNames from 'classnames'

// import TreeLineDown1 from './assets/treeLineDown1.png'
// import TreeLineDown2 from './assets/treeLineDown2.png'
// import TreeLineDownLast from './assets/treeLineDownLast.png'
// import TreeLineTopLast from './assets/treeLine1.png'
// import TreeLineTop2 from './assets/treeLine2.png'
// import TreeLineTopFirst from './assets/treeLineTopLast.png'
// import TreeLineTopLeft from './assets/treeLineLeftTop.png'
// import TreeLineLeftTopFirst from './assets/treeLineLeftTopFirst.png'
// import TreeLineLeftDown from './assets/treeLineLeftDown.png'
// import TreeLineLeftDownLast from './assets/treeLineLeftDownLast.png'
// import TreeLineBottomCenter from './assets/treeLine2.png'
// import TreeLineOnlyOne from './assets/treeLineOnlyOne.png'

const TreeLineTopLast = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABtJREFUOMtjgIAECDXKG+WN8kYs7z8S+IDKAwCg0DWJ9+1vqwAAAABJRU5ErkJggg=='
const TreeLineDown1 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABtJREFUOMtj+I8EPqDxICABQo3yRnmjvJHKAwDzGzWJFvqsYAAAAABJRU5ErkJggg=='
const TreeLineDown2 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABxJREFUOMtjgID6/yDwASsvAUKN8kZ5o7yRygMApWIolRzU7nYAAAAASUVORK5CYII='
const TreeLineDownLast = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABlJREFUOMtjgID6/yDwAY03CkbBKBgFQAAAe5oM1a1+5mYAAAAASUVORK5CYII='
const TreeLineTop2 = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABxJREFUOMtjgIAECDXKG+WN8kY4r/4/CHxA5QEAwZMolaw8DaEAAAAASUVORK5CYII='
const TreeLineTopFirst = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABdJREFUOMtjGAWjYBSMAgSo/w8CH1B5AI6eDNVKPTeWAAAAAElFTkSuQmCC'
const TreeLineTopLeft = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABtJREFUOMtjgIAECDXKG+WN8kYs7z8YPMDGAwAg2il1WWvVOAAAAABJRU5ErkJggg=='
const TreeLineLeftTopFirst = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABZJREFUOMtjGAWjYBSMAhD4DwYPsPEA7dYNtYS4VE0AAAAASUVORK5CYII='
const TreeLineLeftDown = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABtJREFUOMtj+A8GDxggAMFDgIRR3ihvlDeyeQCPVil1Y8Z+KAAAAABJRU5ErkJggg=='
const TreeLineLeftDownLast = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABlJREFUOMtj+A8GBxjAAM4bBaNgFIwCOAAAXv8NdTODBYsAAAAASUVORK5CYII='
const TreeLineBottomCenter = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEUAAAC0tLQrlfMqAAAAAXRSTlMAQObYZgAAABxJREFUOMtjgIAECDXKG+WN8kY4r/4/CHxA5QEAwZMolaw8DaEAAAAASUVORK5CYII='
const TreeLineOnlyOne = 'data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAGQAAABMAQMAAACh04BMAAAABlBMVEVHcEy0tLQR2M60AAAAAXRSTlMAQObYZgAAABdJREFUOMtj+I8EPjCg8kbBKBgFowAIAMlTGcmmcbuAAAAAAElFTkSuQmCC'

const alignConfig = {
  topRight: {
    points: ['bl', 'tr'],
    offset: [ 0, 32],
    // offset: [ 0, 16],
  },
  bottomRight: {
    points: ['tl', 'tr'],
    offset: [0, 0],    
  },
  topLeft: {
    points: ['br', 'tl'],
    offset: [ 0, 32],    
    // offset: [ 0, 16],    
  },
  bottomLeft: {
    points: ['tr', 'bl'],
    offset: [ 0, -29],    
  },
  bottomCenter: {
    points: ['tl', 'bl'],
    offset: [-18, 0],       
  } 
}

class MindTree extends PureComponent {

  constructor(props) {
    super(props);

    this.state = {
      mode: props.mode,
      leafIsOpen: true,
      tree: props.tree,
    }
  }

  componentDidMount() {
    const { bgColor, tree, treeBoxWidth, mode, prefixCls } = this.props
    // const titleDom = document.querySelectorAll(`.${prefixCls}-title`)
    // const treeBox = document.querySelectorAll(`.${prefixCls}-tree-box`)
    if (mode === 'topRight' || mode === 'topLeft') {
      const tree = this.state.tree.slice()
      if (tree && Array.isArray(tree) ) {
        this.setState({
          tree: tree.reverse()
        })
      }
    }
    const titleStyle = {
      ['cursor']: tree ? 'pointer' : 'default'
    }
    const boxStyle = {
      ['width']: treeBoxWidth + 'px'
    }
    // titleDom.forEach((ele) => {
    //   OuiDom.setStyles(ele, titleStyle)
    // })
    OuiDom.setStyles(this.titleDom, titleStyle)
    // treeBox.forEach((ele) => {
    //   OuiDom.setStyles(ele, boxStyle)
    // })
    OuiDom.setStyles(this.leafBox, boxStyle)

    const leafBoxH = OuiDom.outerHeight(this.leafBox)

    const leafTreeStyle = {
      ['width']: treeBoxWidth + 'px',
      ['height']: leafBoxH + 'px',
    }
    // 修正绝对定位的 LeafDom 的宽高度
    OuiDom.setStyles(this.leafTree, leafTreeStyle)
    if (mode) {
      domAlign(this.leafTree, this.titleDom, alignConfig[mode])
    }
  }
  handleTreeChange = (event) => { 
    const { treeChange, prefixCls, mode } = this.props
    const titleDom = event.currentTarget
    const leafDom = titleDom.nextElementSibling.querySelector(`.${prefixCls}-tree-box`)
    const leafDomH = OuiDom.outerHeight(leafDom)
    if (mode === 'topRight' || mode === 'topLeft') {
      if (this.state.leafIsOpen) {
        velocity(leafDom, {
          marginTop: leafDomH + 'px',
          opacity: 0,
        }, {
          duration: 400
        })
      } else {
        velocity(leafDom, {
          marginTop: '0px',
          opacity: 1,
        }, {
          duration: 400
        })      
      }
    } else {
      if (this.state.leafIsOpen) {
        velocity(leafDom, {
          marginTop: -leafDomH + 'px',
          opacity: 0,
        }, {
          duration: 400
        })
      } else {
        velocity(leafDom, {
          marginTop: '0px',
          opacity: 1,
        }, {
          duration: 400
        })          
      }
    }

    this.setState({
      leafIsOpen: !this.state.leafIsOpen
    }, () => {
      if (treeChange && typeof treeChange === 'function') {
        treeChange(this.state.leafIsOpen)
      }
    })

  }

  handleLeafClick = (event, val) => {
    const { leafClick } = this.props
    if (leafClick && typeof leafClick === 'function') {
      leafClick(event, val)
    }
  }

  componentDidUpdate() {
    const {  tree, treeBoxWidth } = this.props
    const titleStyle = {
      ['cursor']: tree ? 'pointer' : 'default'
    }
    const boxStyle = {
      ['width']: treeBoxWidth + 'px'
    }
    OuiDom.setStyles(this.titleDom, titleStyle)
    OuiDom.setStyles(this.leafBox, boxStyle)

    const leafBoxH = OuiDom.outerHeight(this.leafBox)

    const leafTreeStyle = {
      ['width']: treeBoxWidth + 'px',
      ['height']: leafBoxH + 'px',
    }
    // 修正绝对定位的 LeafDom 的宽高度
    OuiDom.setStyles(this.leafTree, leafTreeStyle)
  }
  onMouseEnter = (e) => {
    const { hoverBgColor } = this.props
    const ele = e.currentTarget
    if (hoverBgColor) {
      const hoverStyle = {
        ['background-color']: hoverBgColor
      }
      OuiDom.setStyles(ele, hoverStyle)
    }
  }
  onMouseLeave = (e) => {
    const { hoverBgColor, bgColor } = this.props
    const ele = e.currentTarget
    if (hoverBgColor) {
      const hoverStyle = {
        ['background-color']: bgColor
      }
      OuiDom.setStyles(ele, hoverStyle)        
    } 
  }

  render() {
    const { mode, title, bgColor, prefixCls, className, treeBoxWidth } = this.props
    const tree = this.state.tree
    const treeLast = tree && tree.length > 0 && tree.length - 1
    const mindTreeClassName = classNames({
      [className]: !!className,
      [prefixCls]: true,
    })
    const leafClassName = classNames({
      [`${prefixCls}-tree`]: true,
      [`${prefixCls}-tree-none`]: !(!!tree),
    })
    const leafTextStyle = (topVal) => {
      return {
        backgroundColor: bgColor,
        top: topVal,
      }
    }
    const list = tree && tree.length > 0 && (
      tree.map((item, index) => {
        const nameStyle = item.accessory
          ? { maxWidth:  treeBoxWidth - 100}
          : null
        if (mode === 'topRight') {
          if (tree.length === 1) {
            return (
              <div className={`${prefixCls}-tree-list`} key={index}>
                <span className={`${prefixCls}-tree-list-left`} style={{ marginTop: 45, backgroundImage: `url(${TreeLineOnlyOne})`}}></span>
                <span className={`${prefixCls}-tree-list-text`}
                 style={leafTextStyle('35px')} onClick={e => this.handleLeafClick(e, item)}
                 onMouseEnter={this.onMouseEnter}
                 onMouseLeave={this.onMouseLeave}
                >
                  <span 
                    className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory, [`${prefixCls}-tree-list-addTextOther`]: !item.accessory})}
                    style={nameStyle}
                    title={item.name}
                  >{item.name}</span>
                  { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}
                </span>
              </div>              
            )
          }
          return (
            <div className={`${prefixCls}-tree-list`} key={index}>
              <span className={`${prefixCls}-tree-list-left`} style={{backgroundImage: index === 0 ? `url(${TreeLineTopFirst})` : 
                                                                     index === treeLast ? `url(${TreeLineTopLast})` : `url(${TreeLineTop2})`}}></span>
              <span className={`${prefixCls}-tree-list-text`} 
                style={leafTextStyle('35px')} onClick={e => this.handleLeafClick(e, item)}
                onMouseEnter={this.onMouseEnter}
                onMouseLeave={this.onMouseLeave}
              >
                <span 
                  className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory, [`${prefixCls}-tree-list-addTextOther`]: !item.accessory})}
                  style={nameStyle}
                  title={item.name}
                >{item.name}</span>
                { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}
              </span>
            </div>
          )
        }
        if (mode === 'topLeft') {
          if (tree.length === 1) {
            return (
              <div className={`${prefixCls}-tree-list`} key={index} style={{textAlign: 'right'}}>
                <span className={`${prefixCls}-tree-list-text`} 
                  style={leafTextStyle('35px')} onClick={e => this.handleLeafClick(e, item)}
                  onMouseEnter={this.onMouseEnter}
                  onMouseLeave={this.onMouseLeave}                  
                >
                  <span 
                    className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory, [`${prefixCls}-tree-list-addTextOther`]: !item.accessory})}
                    style={nameStyle}
                    title={item.name}
                  >{item.name}</span>
                  { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}
                </span>
                <span className={`${prefixCls}-tree-list-left`} style={{ marginTop: 45,backgroundImage: `url(${TreeLineOnlyOne})`}}></span>
              </div>              
            )
          }
          return (
            <div className={`${prefixCls}-tree-list`} key={index} style={{textAlign: 'right'}}>
              <span className={`${prefixCls}-tree-list-text`} 
                style={leafTextStyle('35px')} onClick={e => this.handleLeafClick(e, item)}
                onMouseEnter={this.onMouseEnter}
                onMouseLeave={this.onMouseLeave}                
              >
                <span 
                  className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory, [`${prefixCls}-tree-list-addTextOther`]: !item.accessory})}
                  style={nameStyle}
                  title={item.name}
                >{item.name}</span>
                { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}
              </span>
              <span className={`${prefixCls}-tree-list-left`} style={{backgroundImage: index === 0 ? `url(${TreeLineLeftTopFirst})` : 
                                                                     index === treeLast ? `url(${TreeLineTopLast})` : `url(${TreeLineTopLeft})`}}></span>
            </div>
          )          
        }
        if (mode === 'bottomLeft') {
          if (tree.length === 1) {
            return (
              <div className={`${prefixCls}-tree-list`} key={index} style={{textAlign: 'right'}}>
                <span className={`${prefixCls}-tree-list-text`} 
                  style={leafTextStyle('-13px')} onClick={e => this.handleLeafClick(e, item)}
                  onMouseEnter={this.onMouseEnter}
                  onMouseLeave={this.onMouseLeave}                  
                >
                  <span 
                    className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory, [`${prefixCls}-tree-list-addTextOther`]: !item.accessory})}
                    style={nameStyle}
                    title={item.name}
                  >{item.name}</span>
                  { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}
                </span>
                <span className={`${prefixCls}-tree-list-left`} style={{backgroundImage: `url(${TreeLineOnlyOne})`}}></span>
              </div>              
            )
          }
          return (
            <div className={`${prefixCls}-tree-list`} key={index} style={{textAlign: 'right'}}>
              <span className={`${prefixCls}-tree-list-text`} 
                style={leafTextStyle('-13px')} onClick={e => this.handleLeafClick(e, item)}
                onMouseEnter={this.onMouseEnter}
                onMouseLeave={this.onMouseLeave}                
              >
                <span 
                  className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory, [`${prefixCls}-tree-list-addTextOther`]: !item.accessory})}
                  style={nameStyle}
                  title={item.name}
                >{item.name}</span>
                { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}
              </span>
              <span className={`${prefixCls}-tree-list-left`} style={{backgroundImage: index === 0 ? `url(${TreeLineDown1})` : 
                                                                     index === treeLast ? `url(${TreeLineLeftDownLast})` : `url(${TreeLineLeftDown})`}}></span>
            </div>
          )             
        }
        if (mode === 'bottomCenter') {
          return (
            <div className={`${prefixCls}-tree-list`} key={index}>
              <span className={`${prefixCls}-tree-list-left`} style={{backgroundImage: index === 0 ? `url(${TreeLineBottomCenter})` : 
                                                                     index === treeLast ? `url(${TreeLineBottomCenter})` : `url(${TreeLineBottomCenter})`}}></span>
              <span className={`${prefixCls}-tree-list-text`} 
                style={leafTextStyle('35px')} onClick={e => this.handleLeafClick(e, item)}
                onMouseEnter={this.onMouseEnter}
                onMouseLeave={this.onMouseLeave}                
              >
                <span 
                  className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory, [`${prefixCls}-tree-list-addTextOther`]: !item.accessory})}
                  style={nameStyle}
                  title={item.name}
                >{item.name}</span>
                { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}
              </span>
            </div>
          )          
        }
        if (mode === 'bottomRight') {
          if (tree.length === 1) {
            return (
              <div className={`${prefixCls}-tree-list`} key={index}>
                <span className={`${prefixCls}-tree-list-left`} style={{ backgroundImage: `url(${TreeLineOnlyOne})`}}></span>
                <span className={`${prefixCls}-tree-list-text`} 
                  style={leafTextStyle('-13px')} onClick={e => this.handleLeafClick(e, item)}
                  onMouseEnter={this.onMouseEnter}
                  onMouseLeave={this.onMouseLeave}                  
                >
                    <span 
                      className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory, [`${prefixCls}-tree-list-addTextOther`]: !item.accessory})}
                      style={nameStyle}
                      title={item.name}
                    >{item.name}</span>
                    { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}   
                </span>
              </div>  
            )
          }
          return (
            <div className={`${prefixCls}-tree-list`} key={index}>
              <span className={`${prefixCls}-tree-list-left`} style={{ backgroundImage: index === 0 ? `url(${TreeLineDown1})` : 
                                                                    index === treeLast ? `url(${TreeLineDownLast})` : `url(${TreeLineDown2})`}}></span>
              <span className={`${prefixCls}-tree-list-text`} 
                style={leafTextStyle('-13px')} onClick={e => this.handleLeafClick(e, item)}
                onMouseEnter={this.onMouseEnter}
                onMouseLeave={this.onMouseLeave}                
              >
                  <span 
                    className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory, [`${prefixCls}-tree-list-addTextOther`]: !item.accessory})}
                    style={nameStyle}
                    title={item.name}
                  >{item.name}</span>
                  { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}   
              </span>
            </div>  
          )        
        }
        return (
          <div className={`${prefixCls}-tree-list`} key={index}>
            <span className={`${prefixCls}-tree-list-left`} style={{ backgroundImage: index === 0 ? `url(${TreeLineDown1})` : 
                                                                   index === treeLast ? `url(${TreeLineDownLast})` : `url(${TreeLineDown2})`}}></span>
            <span className={`${prefixCls}-tree-list-text`} 
              style={leafTextStyle('-13px')} onClick={e => this.handleLeafClick(e, item)}
              onMouseEnter={this.onMouseEnter}
              onMouseLeave={this.onMouseLeave}              
            >
                <span 
                  className={classNames({[`${prefixCls}-tree-list-addText`]: item.accessory})}
                  style={nameStyle}
                  title={item.name}
                >{item.name}</span>
                { item.accessory && (<span className={`${prefixCls}-tree-list-accessory`}>{item.accessory}</span>)}   
            </span>
          </div>
        )
      })
    )
    return (
      <div className={mindTreeClassName}>
        <div className={`${prefixCls}-title`} 
          style={{backgroundColor: bgColor}} onClick={this.handleTreeChange} ref={titleDom => this.titleDom = titleDom}
          onMouseEnter={this.onMouseEnter}
          onMouseLeave={this.onMouseLeave}
        >
          {
            tree && (tree.length > 0) && (
            <span className={`${prefixCls}-title-decoration`}>
              {
                this.state.leafIsOpen ? (<Icon type="minus-circle" />) : (<Icon type="plus-circle" />)
              }
            </span>
            )
          }
          <span className={`${prefixCls}-title-text`}>
            {title}
          </span>
        </div>
        
        <div className={leafClassName} ref={leafTree => this.leafTree = leafTree}>
          <div className={`${prefixCls}-tree-box`} style={{
            top: (mode === 'bottomLeft' || mode === 'bottomRight') ? '0px' : '-13px',
          }} ref={leafBox => this.leafBox = leafBox}>
            <div 
              className={`${prefixCls}-tree-listBox`} 
              style={{marginTop: mode === 'bottomCenter' ? '0' : '13px'}}
              ref={listBox => this.listBox = listBox}
            >
              {list}
            </div>
          </div>
        </div>
      </div>
    )
  }
}

MindTree.propTypes = {
  mode: PropTypes.oneOf(['topRight', 'bottomRight', 'topLeft', 'bottomLeft', 'bottomCenter']),
  title: PropTypes.string,
  tree: PropTypes.arrayOf(PropTypes.object),
  bgColor: PropTypes.string,
  treeChange: PropTypes.func,
  treeBoxWidth: PropTypes.number,
  leafClick: PropTypes.func,
  prefixCls: PropTypes.string,
  className: PropTypes.string,
  hoverBgColor: PropTypes.string,
}

MindTree.defaultProps = {
  mode: 'bottomRight',
  title: '暂无数据',
  bgColor: '#f5efdc',
  treeBoxWidth: 300,
  prefixCls: 'cr-mind-tree'
}
export default MindTree
